server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/main/java/org/apache/james/webadmin/vault/routes/ExportService.java

```diff
@@ -0,0 +1,118 @@
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.webadmin.vault.routes;
+
+import java.io.IOException;
+import java.io.PipedInputStream;
+import java.io.PipedOutputStream;
+import java.util.stream.Stream;
+
+import javax.inject.Inject;
+
+import org.apache.james.blob.api.BlobId;
+import org.apache.james.blob.api.BlobStore;
+import org.apache.james.blob.sharing.api.BlobSharingMechanism;
+import org.apache.james.core.MailAddress;
+import org.apache.james.core.User;
+import org.apache.james.vault.DeletedMessage;
+import org.apache.james.vault.DeletedMessageVault;
+import org.apache.james.vault.DeletedMessageWithContent;
+import org.apache.james.vault.DeletedMessageZipper;
+import org.apache.james.vault.search.Query;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.fge.lambdas.Throwing;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Schedulers;
+
+class ExportService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ExportService.class);
+
+    private final BlobSharingMechanism blobSharing;
+    private final BlobStore blobStore;
+    private final DeletedMessageZipper zipper;
+    private final DeletedMessageVault vault;
+
+    @Inject
+    ExportService(BlobSharingMechanism blobSharing, BlobStore blobStore, DeletedMessageZipper zipper, DeletedMessageVault vault) {
+        this.blobSharing = blobSharing;
+        this.blobStore = blobStore;
+        this.zipper = zipper;
+        this.vault = vault;
+    }
+
+    Mono<Void> export(User user, Query exportQuery, MailAddress shareeAddress,
+                      Runnable messageToShareCallback) throws IOException {
+        Stream<DeletedMessageWithContent> matchingMessageStream = matchingMessageStream(user, exportQuery)
+            .peek(any -> messageToShareCallback.run());
+
+        PipedInputStream zippedInputStream = zipData(matchingMessageStream);
+        return blobStore.save(zippedInputStream)
+            .flatMap(blobId -> shareTo(user, shareeAddress, blobId))
+            .then();
+    }
+
+    private Stream<DeletedMessageWithContent> matchingMessageStream(User user, Query exportQuery) {
+        return Flux.from(vault.search(user, exportQuery))
+            .publishOn(Schedulers.elastic())
+            .flatMap(deletedMessage -> messageWithContent(user, deletedMessage))
+            .toStream();
```

Comment:
why do you need to use a Stream?

Reply:
because there is a step I don't know the best way to do: transform a Flux<T> to a Mono<Stream<T>>. I can use `Flux.collectToList().map(List::stream)` but it uses one more intermediated  transformation from list to stream which I don't really fancy.